wgengine/userspace.go

```diff
@@ -1313,6 +1327,55 @@ func (e *userspaceEngine) mySelfIPMatchingFamily(dst netaddr.IP) (src netaddr.IP
 	return netaddr.IP{}, errors.New("no self address in netmap matching address family")
 }
 
+func (e *userspaceEngine) sendICMPEcho(ip netaddr.IP, peer *tailcfg.Node, res *ipnstate.PingResult, cb func(*ipnstate.PingResult)) {
```

Comment:
I can probably factor TSMP and ICMP together, but the logic can get a little dense. Happy to go in that direction if it seems worthwhile.

Reply:
is this an echo request or echo reply?